### PR TITLE
Creates mobile commons groups & saves to DB

### PR DESCRIPTION
**Pull request by deadlybutter**
#### What's this PR do?
- Adds some new packages
- Adds new fields to campaign model
- Creates mobile commons groups per campaign
- Saves ID's to model

#### How should this be reviewed?
LOL this is fun. So when testing in Mobile Commons, you cannot simple "delete" a group. You have to rename the group to some random string, then delete it. MoCo reserves the string in namespace even after you delete it.

#### Any background context you want to provide?
Part 2 of this will be exposing this data

#### Relevant tickets
Fixes #

#### Checklist
- [ ] Documentation added for new features/changed endpoints.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

### CODE REVIEW

File: config/locals.js [hunk 1]
```diff
@@ -40,6 +43,44 @@ function cacheCampaign(phoenixCampaign) {
...
 
 /**
+ * Create Doing/Completed Mobile Commons Groups to support Mobile Commons broadcasting.
+ * @see https://github.com/DoSomething/gambit/issues/673
+ */
+function createMobileCommonsGroupsForCampaign(campaignModel) {
+  const campaignScope = campaignModel;
+  const prefix = `env=${process.env.NODE_ENV} campaign_id=${campaignModel._id}`;
+
+  // Migrate old campaign models.
```
Comment (aaronschachter) [inline config/locals.js]: What is this here for?

**Comment by deadlybutter** [comment]
Retested to make sure everything still works
- DB values
- Mobile Commons Groups

File: config/locals.js [hunk 2]
```diff
@@ -40,6 +43,44 @@ function cacheCampaign(phoenixCampaign) {
 }
 
 /**
+ * Create Doing/Completed Mobile Commons Groups to support Mobile Commons broadcasting.
+ * @see https://github.com/DoSomething/gambit/issues/673
+ */
+function createMobileCommonsGroupsForCampaign(campaignModel) {
+  const campaignScope = campaignModel;
+  const prefix = `env=${process.env.NODE_ENV} campaign_id=${campaignModel._id}`;
```
Comment (aaronschachter) [inline config/locals.js]: We'll want to include Campaign Run from https://github.com/DoSomething/phoenix-js/pull/16.

The hope is these Mobile Commons Groups are short-term fix for a more robust subscriber solution for Doing/Completed. But if we're still using this code in a year -- if we want to send a broadcast to all COMPLETED but re-open the campaign, we'll also need to clear the COMPLETED group -- if we add the Current Run ID into the prefix when we create the group, we avoid that scenario.

**Comment by deadlybutter** [comment]
@aaronschachter was already planning on using that fancy new squash & merge button 😉

File: config/locals.js (hunk 2 above)
Comment (deadlybutter) [inline config/locals.js]: @aaronschachter 👍

File: config/locals.js (hunk 1 above)
Comment (deadlybutter) [inline config/locals.js]: Code was breaking on old DB models that didn't have `mobileCommonsGroups.completed`

**Comment by deadlybutter** [comment]
@aaronschachter ready for a second look, heres the MoCo output! https://dosomething.mcommons.com/groups?utf8=%E2%9C%93&filter=env%3Dlocal

File: api/models/Campaign.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ const schema = new mongoose.Schema({
 
   // Properties cached from DS API.
   title: String,
+  run_id: Number,
```
Comment (aaronschachter) [inline api/models/Campaign.js]: None of other Models use a `_id` naming convention -- e.g. `signup.user`, `signup.campaign`.  Let's rename this as `campaign_run`.

If there's a good argument to keep the `_id`, to make it  `campaign_run_id`-- we should change all other properties in other models to keep consistent.

File: api/models/Campaign.js [hunk 4]
```diff
@@ -30,6 +31,12 @@ const schema = new mongoose.Schema({
   msg_menu_signedup: String,
   msg_no_photo_sent: String,
 
+  // Mobile Commons Specific Fields.
+  mobileCommonsGroups: {
```
Comment (aaronschachter) [inline api/models/Campaign.js]: Same here -- all other models are using lowercase/underscore. If we're creating a `mobilecommons` object, I do consider whether `keywords` should be moved into it too.

File: config/locals.js [hunk 5]
```diff
@@ -40,6 +44,46 @@ function cacheCampaign(phoenixCampaign) {
...
+  const prefix = `env=${process.env.NODE_ENV}
+                  campaign_id=${campaignModel._id}
+                  run_id=${campaignModel.run_id}`;
+
+  // Migrate old campaign models.
+  if (!campaignModel.mobileCommonsGroups.doing || !campaignModel.mobileCommonsGroups.completed) {
+    const groups = {
+      doing: '',
+      completed: '',
+    };
```
Comment (aaronschachter) [inline config/locals.js]: Still don't get why we need this?

File: config/locals.js [hunk 6]
```diff
@@ -40,6 +44,46 @@ function cacheCampaign(phoenixCampaign) {
 }
 
 /**
+ * Create Doing/Completed Mobile Commons Groups to support Mobile Commons broadcasting.
+ * @see https://github.com/DoSomething/gambit/issues/673
+ */
+function createMobileCommonsGroupsForCampaign(campaignModel) {
```
Comment (aaronschachter) [inline config/locals.js]: I started on this function before figuring out how to use Mongoose instance method (implemented #675). This would be a perfect candidate to add as an [instance method](http://mongoosejs.com/docs/guide.html#methods) to our `Campaign` model .

File: package.json [hunk 7]
```diff
@@ -42,11 +42,13 @@
     "path": "~0.4.9",
     "q": "^1.0.1",
     "request": "2.34.x",
+    "request-promise-native": "^1.0.3",
```
Comment (aaronschachter) [inline package.json]: I added `supergaent` initially to add Promises support, and because we use it Northstar-JS and Phoenix-JS. Not sure if added another library is anything to flag, but would be curious about @sergii-tkachenko 's opinion

**Comment by aaronschachter** [comment]
Linking this PR to #673

File: config/locals.js (hunk 5 above)
Comment (deadlybutter) [inline config/locals.js]: Ah you're right we don't

File: package.json (hunk 7 above)
Comment (deadlybutter) [inline package.json]: Request by default doesn't have promises, this package adds support for it.

File: api/models/Campaign.js (hunk 3 above)
Comment (aaronschachter) [inline api/models/Campaign.js]: I'm definitely in favor of keeping our naming conventions consistent across platforms, but agree nothing to change today. We do use `currentCampaignRun` in [Phoenix](https://github.com/DoSomething/phoenix-js/blob/7b4c45f7fe31bab7023d2408ce92f89b06d570fa/lib/phoenix-campaign.js#L16) -- my hesitation with `run` is it feels a little too generic, or questionable if it's actual DS Campaign Run. how about `current_run` ?

Comment (deadlybutter) [inline api/models/Campaign.js]: kool 👍

File: api/models/Campaign.js (hunk 4 above)
Comment (deadlybutter) [inline api/models/Campaign.js]: I think a refactor that big should be a different PR

**Comment by deadlybutter** [comment]
@aaronschachter wanna take another look at it? 👀

**Comment by deadlybutter** [comment]
still testing something, disregard review request

**Comment by deadlybutter** [comment]
I PASSED THE TEST. TAKE THAT LINT. THATS RIGHT. I WIN.

**Comment by deadlybutter** [comment]
also @aaronschachter retested everything, its good for review

**Comment by deadlybutter** [comment]
I see it in the app logs..

```
2016-10-24T17:05:58.001166+00:00 app[web.1]:                   campaign_id=2299
2016-10-24T17:05:58.007897+00:00 app[web.1]:                   run_id=6431 status=completed
2016-10-24T17:05:58.007896+00:00 app[web.1]:                   campaign_id=2273
2016-10-24T17:05:58.239370+00:00 app[web.1]: [34mdebug[39m: mobilecommons.createGroup env=thor
2016-10-24T17:05:58.239381+00:00 app[web.1]:                   campaign_id=2900
2016-10-24T17:05:58.250704+00:00 app[web.1]: [34mdebug[39m: mobilecommons.createGroup env=thor
2016-10-24T17:05:58.239382+00:00 app[web.1]:                   run_id=6477 status=completed
2016-10-24T17:05:58.250707+00:00 app[web.1]:                   campaign_id=2938
2016-10-24T17:05:58.250710+00:00 app[web.1]:                   run_id=6487 status=completed
```

Groups on thor created https://dosomething.mcommons.com/groups?utf8=%E2%9C%93&filter=env%3Dthor

Confirmed they're in the DB
screen shot 2016-10-24 at 1 20 09 pm


(Except for one, id 362, which according to the logs didn't load at startup https://gist.github.com/deadlybutter/3656874cf459246ec872dccf1b1134dc )

**Comment by aaronschachter** [comment]
Something's off here. Running locally and the `mobilecommons_groups` aren't appearing -- also staging DB should contain `mobilecommons_groups`, not `mobile_commons_groups`.

File: api/models/Campaign.js [hunk 8]
```diff
@@ -31,8 +35,51 @@ const schema = new mongoose.Schema({
...
 });
 
+/**
+ * For a given campaign, update its mobile commons group id,
+ * based on the mobile commons response.
+ * @param {Campaign} campaign    Campaign model to update.
+ * @param {string} status        Either completed or doing.
+ * @param {string} groupResponse XML string from mobile commons API.
+ */
+function setMobileCommonsGroup(campaign, status, groupResponse) {
```
Comment (aaronschachter) [inline api/models/Campaign.js]: This function doesn't return anything

Comment (aaronschachter) [inline api/models/Campaign.js]: https://github.com/DoSomething/gambit/pull/674/files#r84805483

**Comment by aaronschachter** [comment]
My DB won't update itself upon app start since the Mobile Commons Groups exist already. It'd sure make things easier if the MC `create_group` API endpoint would return the existing Group ID back (or if we could query the `groups` endpoint by name)

**Comment by deadlybutter** [comment]
@aaronschachter we talked about this -- Yeah I totally agree it's be sweet if they had a normal API, but until then this is the way we have to do it :/

File: api/models/Campaign.js [hunk 9]
```diff
@@ -31,8 +35,51 @@ const schema = new mongoose.Schema({
...
+  const prefix = `env=${process.env.NODE_ENV}
+                  campaign_id=${campaign._id}
+                  run_id=${campaign.current_run}`;
+
+  // Create mobile commons group with custom name based on this campaign
+  mobilecommons.createGroup(`${prefix} status=doing`)
+  .then(doingGroup => setMobileCommonsGroup(campaign, 'doing', doingGroup))
+  .then(() => mobilecommons.createGroup(`${prefix} status=completed`))
+  .then(completedGroup => setMobileCommonsGroup(campaign, 'completed', completedGroup))
+  .then(() => campaign.save())
```
Comment (deadlybutter) [inline api/models/Campaign.js]: Campaign is definitely updated, I think this is related to the problem we talked about below with you not being able to pull data.

**Comment by deadlybutter** [comment]
Also remember you can't delete the group, you have to rename it then delete it! :facepalm:

File: api/models/Campaign.js [hunk 10]
```diff
@@ -31,8 +35,51 @@ const schema = new mongoose.Schema({
...
+/**
+ * For a given campaign, update its mobile commons group id,
+ * based on the mobile commons response.
+ * @param {Campaign} campaign    Campaign model to update.
+ * @param {string} status        Either completed or doing.
+ * @param {string} groupResponse XML string from mobile commons API.
+ */
+function setMobileCommonsGroup(campaign, status, groupResponse) {
+  const scope = campaign;
+  const parsedGroup = JSON.parse(parser.toJson(groupResponse));
```
Comment (aaronschachter) [inline api/models/Campaign.js]: @deadlybutter would it be cleaner to move parsing the XML response into the `mobileCommons.createGroup` function?

Comment (deadlybutter) [inline api/models/Campaign.js]: I originally did that but for some reason the parsed response went back to XML when the caller function tried using the response. I got annoyed & moved it haha

Comment (aaronschachter) [inline api/models/Campaign.js]: My 0.02 is it would be a lot cleaner to parse the JSON in `lib/mobilecommons.js` -- especially since we'll want to parse there too for a `getGroups` function last resort in https://github.com/DoSomething/gambit/issues/673#issuecomment-256251277
